bundles/persistence/org.openhab.persistence.dynamodb/java/org/openhab/persistence/dynamodb/internal/DynamoDBPersistenceService.java

```diff
@@ -8,67 +8,147 @@
 
 /**
  * This is the implementation of the DynamoDB {@link PersistenceService}. It persists item values
  * using the <a href="https://aws.amazon.com/dynamodb/">Amazon DynamoDB</a> database. The states (
- * {@link State}) of an {@link Item} are persisted in a time series with names equal to the name of
- * the item. All values are stored using integers or doubles, {@link OnOffType} and
- * {@link OpenClosedType} are stored using 0 or 1.
+ * {@link State}) of an {@link Item} are persisted in DynamoDB tables.
  *
- * The default database name is "openhab"
+ * The service creates tables automatically, one for numbers, and one for strings.
+ *
+ * @see AbstractDynamoDBItem.fromState for details how different items are persisted
  *
  * @author Sami Salonen
  *
  */
-public class DynamoDBPersistenceService implements QueryablePersistenceService {
+public class DynamoDBPersistenceService extends AbstractBufferedPersistenceService<DynamoDBItem<?>>
+        implements QueryablePersistenceService {
+
+    private class ExponentialBackoffRetry implements Runnable {
+        private int retry;
+        private Map<String, List<WriteRequest>> unprocessedItems;
+        private Exception lastException;
+
+        public ExponentialBackoffRetry(Map<String, List<WriteRequest>> unprocessedItems) {
+            this.unprocessedItems = unprocessedItems;
+        }
+
+        @Override
+        public void run() {
+            logger.debug("Error storing object to dynamo, unprocessed items: {}. Retrying with exponential back-off",
+                    unprocessedItems);
+            lastException = null;
+            while (!unprocessedItems.isEmpty() && retry < WAIT_MILLIS_IN_RETRIES.length) {
+                if (!sleep()) {
+                    // Interrupted
+                    return;
+                }
+                retry++;
+                try {
+                    BatchWriteItemOutcome outcome = DynamoDBPersistenceService.this.db.getDynamoDB()
+                            .batchWriteItemUnprocessed(unprocessedItems);
+                    unprocessedItems = outcome.getUnprocessedItems();
+                    lastException = null;
+                } catch (AmazonServiceException e) {
+                    if (e instanceof ResourceNotFoundException) {
+                        logger.debug(
+                                "DynamoDB query raised unexpected exception: {}. This might happen if table was recently created",
+                                e.getMessage());
+                    } else {
+                        logger.debug("DynamoDB query raised unexpected exception: {}.", e.getMessage());
+                    }
+                    lastException = e;
+                    continue;
+                }
+            }
+            if (unprocessedItems.isEmpty()) {
+                logger.debug("After {} retries successfully wrote all unprocessed items", retry);
+            } else {
+                logger.error(
```

Comment:
This should be a WARN instead of an ERROR.

Reply:
Fixed in 0f3cf54